XWIKI-20843: Application Panel "More Applications" dropdown is not accessible

Sereza7 opened the pull request:
**Jira:** https://jira.xwiki.org/browse/XWIKI-20843
**PR Changes:**
* Replaced the anchor for _more applications_ in the Applications panel with a button.
* Updated style
* Added semantics to the disclosure

**View:**
Before PR:
![20843-Before](https://user-images.githubusercontent.com/28761965/234214228-08ec2031-0025-4a8d-9652-fdca3fc5cf8a.png)
After PR:
![20843-After](https://user-images.githubusercontent.com/28761965/234234934-790470f1-9dda-4c3f-83ec-85efdb241d73.png)
The button has the same style when the disclosure is closed.

**Demo:** (made before the last changes on style, but still very similar even on style)
https://up1.xwikisas.com/#NdrhBiSPlXXJNnwWHWfPtA

**Notes:**
* This section is visible only by admins and not regular users.
* For regular users, the section appears and then gets hidden. It was not very noticeable but with a blue button it's way easier to notice this behavior. I couldn't find the code responsible for hiding the section so I didn't fix this. 
https://up1.xwikisas.com/#504lX2oFtkkPm8M9w-L7XQ

Sereza7 commented:
@michitux 
> I think the button is quite a style change and imho it has way too much visual weight. 

Okay for the style change, I'll reduce the weight 'artificially' with specific CSS rules to make it more similar to a link

> Wouldn't this kind of element be a perfect application of the [details](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/details) HTML element which should have very good built-in accessibility that wouldn't even need the JavaScript code for the toggle?

As seen in this [detailed blogpost from an accessibility specialist](https://www.scottohara.me//blog/2022/09/12/details-summary.html) published in September 2022, there is lots of pros to using details, but also lots of cons especially concerning assistive tech support. I checked out [the ARIA Authoring Practice Guide](https://www.w3.org/WAI/ARIA/apg/patterns/) and they don't mention \<details> in any of their patterns.

> 
> Learn how the \<summary> element is exposed to AT:
> 
> Depending on the screen reader and browser pairing you may be using, the \<summary> element would have been exposed in one of the following ways:
> 
> "Disclosure Triangle" with Narrator, VoiceOver and TalkBack when paired with Edge/Chrome.
> "Button" with NVDA when paired with Firefox, Edge or Chrome.
> "Summary" with VoiceOver when paired with Firefox or Safari.
> Bugged behavior or no role announced with TalkBack with Firefox or iOS VO with Safari, respectively.

From [Scott O'Hara's blog](https://www.scottohara.me//blog/2022/09/12/details-summary.html)

So I decided to go with the recommended solution from ARIA APG and [just use a button to toggle the dropdown](https://www.w3.org/WAI/ARIA/apg/patterns/disclosure/examples/disclosure-faq/). Using this standard pattern ensures that we shouldn't be too affected by changes in the ARIA specs and shouldn't have too much technical debt.

I can't say for certain what solution will create the least technical debt (maybe in a couple months all web browsers are going to have good support for \<details> and it'll be the new recommendation!... but it's been already a few years :S). It seems to me that using a button should be fine on short and middle time scale.
I think both would be correct on a WCAG standpoint (as in, we do fulfill our part of the job in making xwiki content accessible), but one is definitely better for assistive technologies users.

At last, if the button pattern does become obsolete on the long term, it shouldn't be too difficult to replace.

michitux commented:
Okay, I didn't expect these problems with `<details>`. I agree with using a button to toggle the links.

Regarding the appearance, I'm not totally opposed to making this a button visually, but then please a button more like the "Create" button in the content menu and not a blue button.

You could also open a proposal to get some more opinions.

Sereza7 commented:
@michitux 
I created a forum proposal for this topic :) 
https://forum.xwiki.org/t/ui-application-panel-dropdown-toggler/12602

Sereza7 commented:
## PR Changes
* Updated style to keep the link presentation

## View
![20843-NonAccessibilityIcebergView](https://github.com/xwiki/xwiki-platform/assets/28761965/ab6775ad-3414-4f88-a01c-c309ffe45304)
![20843-AccesssibleIcebergView](https://github.com/xwiki/xwiki-platform/assets/28761965/dd5d1c30-c709-4518-8435-b96ab12083b1)


## TODO

Clean up the javascript extension, there is some code that's left unused as of now...

Sereza7 commented:
## PR Changes
* Updated style
* Removed unused style classes

## View
Iceberg - Collapsed menu (hover)
![20843-IcebergCollapsed](https://github.com/xwiki/xwiki-platform/assets/28761965/c3b7233d-52f3-45b2-bc35-69858c02bd75)
Iceberg - Expanded menu 
![20843-IcebergExpanded](https://github.com/xwiki/xwiki-platform/assets/28761965/ea36df72-7ac6-4e70-8b27-fde754c278c0)
Darkly - Expanded menu
![20843-Darkly](https://github.com/xwiki/xwiki-platform/assets/28761965/d0b1c1c0-03a0-47a7-91aa-5c23a51b3da2)

michitux commented on xwiki-platform-core/xwiki-platform-panels/xwiki-platform-panels-ui/src/main/resources/Panels/Applications.xml:
```diff
@@ -553,13 +550,34 @@ require(['scriptaculous/effects'], function() {
-.applicationPanelMoreContainer li a {
-  font-size: 90%;
-  line-height: 16px;
+.applicationPanelMoreButton .application-label{
```
I think we normally put a space before `{`.

Sereza7 replied:
Fixed in https://github.com/xwiki/xwiki-platform/pull/2157/commits/96471a12f760ccb90ac33bfa00f3c66e98b97aef

michitux commented on xwiki-platform-core/xwiki-platform-panels/xwiki-platform-panels-ui/src/main/resources/Panels/Applications.xml:
```diff
@@ -553,13 +550,34 @@ require(['scriptaculous/effects'], function() {
+  padding: 0;
+}
+
+.panel-width-Small .applicationPanelMoreButton .application-label{
```
Same as above, missing space before `{`.

Sereza7 replied:
Fixed in https://github.com/xwiki/xwiki-platform/pull/2157/commits/96471a12f760ccb90ac33bfa00f3c66e98b97aef

michitux commented on xwiki-platform-core/xwiki-platform-panels/xwiki-platform-panels-ui/src/main/resources/Panels/Applications.xml:
```diff
@@ -232,21 +232,22 @@
 #panelheader($services.localization.render('panels.applications.title'))
 
 {{html}}
-&lt;ul class="applicationsPanel nav nav-pills nav-stacked"&gt;
```
Why did you remove these classes?

Sereza7 replied:
Because removing them did not change significantly the presentation, and they brought an unexpected style from bootstrap 3 that would make the color match badly on some color themes. 
On Darkly, the child anchors were just white instead of being green like any other anchor in the panels, and almost anywhere on the page vv
![20843-removeBootstrapNav](https://github.com/xwiki/xwiki-platform/assets/28761965/92824037-2367-41d5-84b8-a9271e29efe0)
Moreover I think this reduces the dependency on Bootstrap 3, which is a good thing if we want to move on from it someday. :)

michitux replied:
Looking at this screenshot, the indentation of the "More applications" seems changed. Previously, this was aligned with the other entries like "Dashboard". Why has this alignment changed?

michitux reviewed (changes requested):
Some small comments. Is it correct that this shouldn't change the appearance now, at least not when hovering/focusing the button?

Sereza7 commented:
@michitux The button had a background style change for :hover, however I did forget about the :focus pseudo class.
This is fixed in https://github.com/xwiki/xwiki-platform/pull/2157/commits/01170f804ad3240cb42327161f4b2fe3eccd5310 .
Here is an updated demo of this PR: https://up1.xwikisas.com/#Z82BiviqdKeiKakqsd625w

michitux commented on xwiki-platform-core/xwiki-platform-panels/xwiki-platform-panels-ui/src/main/resources/Panels/Applications.xml:
```diff
@@ -232,21 +232,22 @@
   &lt;li&gt;
-    &lt;a href="$xwiki.getURL($services.model.createDocumentReference('', 'XWiki', 'XWikiPreferences'), 'admin',
-        'editor=globaladmin&amp;section=XWiki.AddExtensions')" class="applicationPanelMoreButton" title="$services.localization.render('panels.applications.more')"&gt;
+    #set ($panelId = $doc.getXDOM().getIdGenerator().generateUniqueId('applicationPanelMoreContainer',''))
```
I hadn't noticed that before. This seems wrong. `$doc.getXDOM()` returns a clone of a cached XDOM instance, I don't think it makes sense to use its id generator.

Sereza7 replied:
@michitux 
I'm not sure this is possible to keep the XDOM object through different calls of this panel.
I can't find how to access a proper IdGenerator here.
Instead, we could append a random alphanumeric string at the end of the string to make it 'unique'.

Change proposed in https://github.com/xwiki/xwiki-platform/pull/2157/commits/926c9109b2023237c74aa4d740fe489dcdd7d2e3
![20843-UniqueID](https://github.com/xwiki/xwiki-platform/assets/28761965/800dcbd4-41bc-44a3-a989-dc104dfef04f)
Note that this is very similar to the solution used before (10 alphanumeric instead of a number from 1 to 100, but the concept is the same).
